subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/processors/RunSortedBySizeTestClassProcessor.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.processors;
+
+import org.gradle.api.internal.tasks.testing.TestClassProcessor;
+import org.gradle.api.internal.tasks.testing.TestClassRunInfo;
+import org.gradle.api.internal.tasks.testing.TestResultProcessor;
+
+import java.util.Comparator;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+/**
+ * This processor uses test class duration information to resort test classes by
+ * duration in descending order before passing them on for processing.
+ *
+ * Tests processed without duration information are at the end behind known classes.
+ * If no durations at all are known beforehand, this will not change the test order.
+ */
+public class RunSortedBySizeTestClassProcessor implements TestClassProcessor {
+    private final TestClassProcessor delegate;
+    private final Map<String, Long> durationByTestClassName;
+
+    private final SortedSet<TestClassRunInfo> knownTestClasses = new TreeSet<TestClassRunInfo>(new Comparator<TestClassRunInfo>() {
```

Comment:
Why is that?
TreeSet consumes more memory than ArrayList, and it is slower.

Reply:
You had got me thinking about the List versus Set issue here (since I had not been satisfied with the mix myself). The processor I modelled this on (the RunPreviousFailedFirstTestClassProcessor) also uses sets, which from an API point of view makes more sense than the lists, since we don't want duplicates.

As far as speed is concerned, the costly operation here would seem to be the sorting, which we have to do either way. Even if we talk about tens of thousands of tests, I don't believe the memory footprint difference is something to worry about here.

Reply:
Even though both memory and speed should not be an issue, I don't see reasons to write code in such a way that it is inherently less efficient.

Note: sorting of `ArrayList` is way faster than inserting the elements one by one to `TreeSet`.

> which from an API point of view makes more sense than the lists, since we don't want duplicates.

It does not. If there are duplicates, then the same test might be silently assigned to a different executors.
Note: there are 3 or so layers (e.g. `failed tests first`, `slowest tests first`, and so on). Does it make sense to `deduplicate` the set of tests 3 times? I doubt so.
That is why I would suggest going with `ArrayList` for both `known duration` and `unknown duration` tests.
Tests with known duration are to be sorted. Tests with unknown duration should be shuffled (==to prevent ill case coming again and again).

If deduplication is really needed, it should be handled as one more layer.

Reply:
While we may disagree about some issues, I do like your interpretation of the `TestClassProcessor` interface in that it admittedly does not imply the need to always guarantee uniqueness; and that it should hence be legitimate to maintain the status quo by extracting this behaviour from the `RunPreviousFailedFirstTestClassProcessor` to an outer layer. And using a list in the sorting processor is simpler, no doubt :)

I have not added shuffling - please see my reply to your previous comment on that.